internal/graph/check.go

```diff
@@ -805,64 +848,181 @@ func buildTupleKeyConditionFilter(ctx context.Context, reqCtx *structpb.Struct,
+			if !errors.Is(err, storage.ErrIteratorDone) && !errors.Is(err, context.Canceled) {
+				trySendUsersetsError(ctx, err, usersetsChan)
 			}
-			telemetry.TraceError(span, err)
-			return nil, err
+			break ProducerLoop
 		}
 
-		object, relation := tuple.SplitObjectRelation(t.GetUser())
-		objectType, objectID := tuple.SplitObject(object)
-		terminalRelations := typesys.GetTerminalRelations(objectType, relation, reqUserType)
-		// the terminalRelations is expected to be 1 (as we checked earlier in typesys.UsersetCanFastPath)
-		if len(terminalRelations) != 1 {
-			return nil, fmt.Errorf("expected exactly one terminal relation for fast path, received %d", len(terminalRelations))
+		objectRel, objectID, err := usersetDetails(t)
+		if err != nil {
+			trySendUsersetsError(ctx, err, usersetsChan)
+			break ProducerLoop
 		}
-		computedRelation := terminalRelations[0]
-		objectRel := tuple.ToObjectRelationString(objectType, computedRelation)
+
 		if _, ok := usersetsMap[objectRel]; !ok {
+			if len(usersetsMap) >= 1 {
+				// Flush results from other usersets so that we can start processing those now.
+				// The assumption (which may not be true, but we don't care) is that the datastore gives us usersets in order.
+				trySendUsersetsAndDeleteFromMap(ctx, usersetsMap, usersetsChan)
+			}
 			usersetsMap[objectRel] = storage.NewSortedSet()
 		}
+
 		usersetsMap[objectRel].Add(objectID)
+		if usersetsMap[objectRel].Size() > int(c.usersetBatchSize) {
+			select {
+			case <-ctx.Done():
+				break ProducerLoop
+			case usersetsChan <- usersetsChannelType{
+				objectRelation: objectRel,
+				objectIDs:      usersetsMap[objectRel],
+			}:
+				// flushed current batch, restarting objectRel entry
+				usersetsMap[objectRel] = storage.NewSortedSet()
+			}
+		}
 	}
 
-	// Next, for all the ObjectRelation, compare the associated objectIDs
-	// to the users associated objects
-	// all of this can likely bee its own function
-	handlers := make([]CheckHandlerFunc, 0, len(usersetsMap))
-	for objectRel, objectIDs := range usersetsMap {
-		handler := c.buildCheckAssociatedObjects(req, objectRel, objectIDs)
-		handlers = append(handlers, handler)
-	}
-	resp, err := union(ctx, c.concurrencyLimit, handlers...)
-	if err != nil {
-		telemetry.TraceError(span, err)
-		return nil, err
-	}
+	trySendUsersetsAndDeleteFromMap(ctx, usersetsMap, usersetsChan)
+}
 
-	resp.GetResolutionMetadata().DatastoreQueryCount++
+func trySendUsersetsError(ctx context.Context, err error, errorChan chan usersetsChannelType) {
+	select {
+	case <-ctx.Done():
+	case errorChan <- usersetsChannelType{err: err}:
+	}
+}
 
-	return resp, nil
+func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMapType, usersetsChan chan usersetsChannelType) {
+	for k, v := range usersetsMap {
+		select {
+		case <-ctx.Done():
+			return
+		case usersetsChan <- usersetsChannelType{
+			objectRelation: k,
+			objectIDs:      v,
+		}:
+			delete(usersetsMap, k)
```

Comment:
curious why do we need to delete individual item vs. using a new usersetMap

Reply:
I cleaned it up a bit without seeing this comment. But we clear it in order to keep track of `seen` objectRelationships and flushing when we observe a new one (based on the assumption that they will come sorted).